File: binding/src/libqthreadpool.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qthreadpool.html
pub const qthreadpool = struct {

    /// New constructs a new QThreadPool object.
    pub fn New() QtC.QThreadPool {
        return qtc.QThreadPool_new();
    }


    /// New2 constructs a new QThreadPool object.
    pub fn New2(parent: ?*anyopaque) QtC.QThreadPool {
        return qtc.QThreadPool_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QThreadPool_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qthreadpool.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn GlobalInstance() QtC.QThreadPool {
        return qtc.QThreadPool_GlobalInstance();
    }

    pub fn Start(self: ?*anyopaque, runnable: ?*anyopaque) void {
        qtc.QThreadPool_Start(@ptrCast(self), @ptrCast(runnable));
    }

    pub fn TryStart(self: ?*anyopaque, runnable: ?*anyopaque) bool {
        return qtc.QThreadPool_TryStart(@ptrCast(self), @ptrCast(runnable));
    }

    pub fn StartOnReservedThread(self: ?*anyopaque, runnable: ?*anyopaque) void {
        qtc.QThreadPool_StartOnReservedThread(@ptrCast(self), @ptrCast(runnable));
    }

    pub fn ExpiryTimeout(self: ?*anyopaque, ) i32 {
        return qtc.QThreadPool_ExpiryTimeout(@ptrCast(self));
    }

    pub fn SetExpiryTimeout(self: ?*anyopaque, expiryTimeout: i32) void {
        qtc.QThreadPool_SetExpiryTimeout(@ptrCast(self), expiryTimeout);
    }

    pub fn MaxThreadCount(self: ?*anyopaque, ) i32 {
        return qtc.QThreadPool_MaxThreadCount(@ptrCast(self));
    }

    pub fn SetMaxThreadCount(self: ?*anyopaque, maxThreadCount: i32) void {
        qtc.QThreadPool_SetMaxThreadCount(@ptrCast(self), maxThreadCount);
    }

    pub fn ActiveThreadCount(self: ?*anyopaque, ) i32 {
        return qtc.QThreadPool_ActiveThreadCount(@ptrCast(self));
    }

    pub fn SetStackSize(self: ?*anyopaque, stackSize: u32) void {
        qtc.QThreadPool_SetStackSize(@ptrCast(self), stackSize);
    }

    pub fn StackSize(self: ?*anyopaque, ) u32 {
        return qtc.QThreadPool_StackSize(@ptrCast(self));
    }

    pub fn SetThreadPriority(self: ?*anyopaque, priority: i32) void {
        qtc.QThreadPool_SetThreadPriority(@ptrCast(self), @intCast(priority));
    }

    pub fn ThreadPriority(self: ?*anyopaque, ) i32 {
        return qtc.QThreadPool_ThreadPriority(@ptrCast(self));
    }

    pub fn ReserveThread(self: ?*anyopaque, ) void {
        qtc.QThreadPool_ReserveThread(@ptrCast(self));
    }

    pub fn ReleaseThread(self: ?*anyopaque, ) void {
        qtc.QThreadPool_ReleaseThread(@ptrCast(self));
    }

    pub fn WaitForDone(self: ?*anyopaque, msecs: i32) bool {
        return qtc.QThreadPool_WaitForDone(@ptrCast(self), msecs);
    }

    pub fn WaitForDone2(self: ?*anyopaque, ) bool {
        return qtc.QThreadPool_WaitForDone2(@ptrCast(self));
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QThreadPool_Clear(@ptrCast(self));
    }

    pub fn Contains(self: ?*anyopaque, thread: ?*anyopaque) bool {
        return qtc.QThreadPool_Contains(@ptrCast(self), @ptrCast(thread));
    }

    pub fn TryTake(self: ?*anyopaque, runnable: ?*anyopaque) bool {
        return qtc.QThreadPool_TryTake(@ptrCast(self), @ptrCast(runnable));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QThreadPool_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qthreadpool.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QThreadPool_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qthreadpool.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Start2(self: ?*anyopaque, runnable: ?*anyopaque, priority: i32) void {
        qtc.QThreadPool_Start2(@ptrCast(self), @ptrCast(runnable), priority);
    }

    pub fn WaitForDone1(self: ?*anyopaque, deadline: QtC.QDeadlineTimer) bool {
        return qtc.QThreadPool_WaitForDone1(@ptrCast(self), @ptrCast(deadline));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QThreadPool_Delete(@ptrCast(self));
    }
};

File: binding/src/libqthreadpool.cpp
#include <QDeadlineTimer>
#include <QObject>
#include <QRunnable>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QThread>
#include <QThreadPool>
#include <qthreadpool.h>
#include "libqthreadpool.h"
#include "libqthreadpool.hxx"

QThreadPool* QThreadPool_new() {
	 return new QThreadPool();
}

QThreadPool* QThreadPool_new2(QObject* parent) {
	 return new QThreadPool(parent);
}

libqt_string QThreadPool_Tr(const char* s) {
	QString _ret = QThreadPool::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QThreadPool* QThreadPool_GlobalInstance() {
	return QThreadPool::globalInstance();
}

void QThreadPool_Start(QThreadPool* self, QRunnable* runnable) {
	self->start(runnable);
}

bool QThreadPool_TryStart(QThreadPool* self, QRunnable* runnable) {
	return self->tryStart(runnable);
}

void QThreadPool_StartOnReservedThread(QThreadPool* self, QRunnable* runnable) {
	self->startOnReservedThread(runnable);
}

int QThreadPool_ExpiryTimeout(const QThreadPool* self) {
	return self->expiryTimeout();
}

void QThreadPool_SetExpiryTimeout(QThreadPool* self, int expiryTimeout) {
	self->setExpiryTimeout(expiryTimeout);
}

int QThreadPool_MaxThreadCount(const QThreadPool* self) {
	return self->maxThreadCount();
}

void QThreadPool_SetMaxThreadCount(QThreadPool* self, int maxThreadCount) {
	self->setMaxThreadCount(maxThreadCount);
}

int QThreadPool_ActiveThreadCount(const QThreadPool* self) {
	return self->activeThreadCount();
}

void QThreadPool_SetStackSize(QThreadPool* self, unsigned int stackSize) {
	self->setStackSize(stackSize);
}

unsigned int QThreadPool_StackSize(const QThreadPool* self) {
	return self->stackSize();
}

void QThreadPool_SetThreadPriority(QThreadPool* self, int priority) {
	self->setThreadPriority(static_cast<QThread::Priority>(priority));
}

int QThreadPool_ThreadPriority(const QThreadPool* self) {
	return self->threadPriority();
}

void QThreadPool_ReserveThread(QThreadPool* self) {
	self->reserveThread();
}

void QThreadPool_ReleaseThread(QThreadPool* self) {
	self->releaseThread();
}

bool QThreadPool_WaitForDone(QThreadPool* self, int msecs) {
	return self->waitForDone(msecs);
}

bool QThreadPool_WaitForDone2(QThreadPool* self) {
	return self->waitForDone();
}

void QThreadPool_Clear(QThreadPool* self) {
	self->clear();
}

bool QThreadPool_Contains(const QThreadPool* self, const QThread* thread) {
	return self->contains(thread);
}

bool QThreadPool_TryTake(QThreadPool* self, QRunnable* runnable) {
	return self->tryTake(runnable);
}

libqt_string QThreadPool_Tr2(const char* s, const char* c) {
	QString _ret = QThreadPool::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QThreadPool_Tr3(const char* s, const char* c, int n) {
	QString _ret = QThreadPool::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QThreadPool_Start2(QThreadPool* self, QRunnable* runnable, int priority) {
	self->start(runnable, priority);
}

bool QThreadPool_WaitForDone1(QThreadPool* self, QDeadlineTimer* deadline) {
	return self->waitForDone(*deadline);
}

void QThreadPool_Delete(QThreadPool* self) {
    delete self;
}

